### ISSUE TITLE
build: support debug mode with symbols for CGo

### ISSUE THREAD

**Issue by arjunravinarayan** [cockroachdb/cockroach#19651]
Currently, using `go tool pprof` does not show symbols for C++ function calls out of the box.

in order to turn them on, I took the following steps (It is not clear which of the steps are strictly necessary):

0. Add `_ "github.com/ianlancetaylor/cgosymbolizer"` to your import if
   this is a benchmark you are `pprof`ing.

1. add `,-g` to common.mk's line ~410 after the `-msse3` in `$(ROCKSDB_DIR)/Makefile`.

2. Run `make clean && CXXFLAGS=-g make` so that everything is rebuilt.

3a. `go tool pprof` on a linux binary will take a *lot* longer. Be patient.

3b. You can use `go tool pprof -symbolize=none` if you cannot wait,
   but this doesn't demangle C++ symbols, so your pprof output looks
   bad.

It would be nice if we could automate parts of this process, especially the builder flags with some `DEBUG` mode.

**Comment by a-robinson** [comment]
I assume not, but are any pprof profile types other than CPU supported across the cgo boundary? Block profiles in particular would be pretty snazzy.

**Comment by bdarnell** [comment]
Not in a unified way, but we can get C++-side heap profiles from `jemalloc`. 

[Abseil's mutex](https://abseil.io/about/design/mutex#extra-features) has partial support for mutex profiling. Changing rocksdb to use abseil's mutex (instead of pthread's) and then exporting this profile information would be useful but a substantial project.

We might be able to generate a `cgosymbolizer` import into one of the `zcgo_flags.go` files to eliminate the need to manually modify things (putting this import in cgo-using packages makes more sense than its current location in package `cli`).

**Comment by tschottdorf** [comment]
@bdarnell how do you feel about compiling everything (in particular rocksdb) with debug symbols? I feel most strongly about this for `make build` and less strongly about it for the release builds (though there too it would be useful when debugging customer issues which are often housed in the RocksDB code).

My understanding is that the argument against including debug symbols is the larger size of the binary which may affect caching (?). In that case, we could strip the release binary but separate out the debug symbols (`strip --debug-only -o something.dwf` or a similar incantation), though we'd then have to figure out how to marry this back into pprof. Might be easier to keep both the stripped and unstripped binary around so that we can just download the unstripped one when we need the symbols. (I hope this is how this works, am a little out of my depth here).

**Comment by bdarnell** [comment]
> how do you feel about compiling everything (in particular rocksdb) with debug symbols?

I'm generally in favor.

> My understanding is that the argument against including debug symbols is the larger size of the binary which may affect caching (?).

AIUI, the biggest drawback to doing this by default is that because all the object and binary files are larger, the increased I/O can slow down the build. The size of the final binary is a secondary concern, and even then I think it's more about conserving download bandwidth and disk space than memory use during execution (since the debug symbols don't need to be paged in). 

We should time the build with and without debug symbols, both from clean and after touching a single file (debug symbols disproportionately impact link time, which is run even in incremental builds). If it doesn't make a significant difference, we should just turn it on all the time. If it does slow the build down, we might want to provide an opt-out.

**Comment by arjunravinarayan** [comment]
@tschottdorf FWIW, there's no concerns about cache behavior in execution with debug symbols. It's just the straightforward things that @bdarnell said - size of the executable in downloads, build times, etc.

**Comment by tschottdorf** [comment]
@bdarnell so what I missed earlier is that `make build` already doesn't strip anything, so nothing will change there except that the addition of RocksDB debug symbols will blow up the file.

For `release-linux-gnu` the TL;DR is that the binary goes from 53mb to 185mb and `make build` after changing the `cli` package takes ~12s instead of ~7. As a comparison, the stock `make build` (which has symbols) is ~91mb, so anecdotally the addition of the RocksDB symbols adds ~100mb.

Note that this is omitting both the `-s` and `-w` link flag. I'm unable to find the documentation for `-w`, but I'm trusting @benesch that it needs to be omitted along with `-s`. That assumption definitely needs a check, though. I'm also curious about the size inflation. `nm cockroach | wc -l` on stock master gives ~86k lines. With RocksDB it went up to ~89 but the binary gains 100mb. Not sure if that's expected.

I used the below script on my 24 core gceworker.

```bash
#!/bin/bash

set -euo pipefail

for branch in master build/debugsymbols; do
  git checkout "${branch}"
  make -j 1 unsafe-clean
  name="$(git symbolic-ref -q HEAD)"
  time ./build/builder.sh make build TYPE=release-linux-gnu
  echo "^-- full build on $name GREPME"
  echo >> pkg/cli/start.go # touch is not enough
  time ./build/builder.sh make build TYPE=release-linux-gnu
  echo "^-- single Go file changed build $name GREPME"
  time ./build/builder.sh make build TYPE=release-linux-gnu
  echo "^-- nothing changed build $name GREPME"
  ls -lah cockroach*
  echo "#symbols: $(nm cockroach* | wc -l)"
  file cockroach*
  echo "^-- file info for $name GREPME"
done
```

**Comment by benesch** [comment]
Haven't actually tested to see what happens with `-w`, but it seemed relevant:

```
$ go tool link
--- 8< ---
  -w	disable DWARF generation
```

**Comment by benesch** [comment]
Also, yeah, debug information can be huge. https://github.com/google/bloaty if you want to investigate further.

**Comment by benesch** [comment]
Here's our typical make build binary:

```
     VM SIZE                         FILE SIZE
 --------------                   --------------
  48.6%  25.8Mi .text              25.8Mi  28.4%
   0.0%       0 .debug_info        16.0Mi  17.6%
  19.5%  10.3Mi .gopclntab         10.3Mi  11.4%
  15.6%  8.29Mi .rodata            8.29Mi   9.1%
   0.0%       0 .strtab            5.67Mi   6.2%
  10.0%  5.30Mi .noptrdata         5.30Mi   5.8%
   0.0%       0 .debug_loc         2.75Mi   3.0%
   0.0%       0 .debug_line        2.65Mi   2.9%
   0.0%       0 .debug_pubtypes    2.59Mi   2.8%
   0.0%       0 .debug_frame       2.06Mi   2.3%
   0.0%       0 .debug_pubnames    2.02Mi   2.2%
   0.0%       0 .symtab            1.98Mi   2.2%
   0.0%       0 .debug_ranges      1.32Mi   1.5%
   2.0%  1.07Mi .dynstr            1.07Mi   1.2%
   0.0%       0 .debug_str          920Ki   1.0%
   0.9%   488Ki [37 Others]         585Ki   0.6%
   1.0%   554Ki .gcc_except_table   554Ki   0.6%
   0.7%   405Ki .eh_frame           405Ki   0.4%
   0.6%   346Ki .bss                    0   0.0%
   0.6%   328Ki .dynsym             328Ki   0.4%
   0.4%   231Ki .data               231Ki   0.2%
 100.0%  53.1Mi TOTAL              90.8Mi 100.0%
```

Our typical release binary:

```
$ bloaty cockroach-linux-2.6.32-gnu-amd64 
     VM SIZE                         FILE SIZE
 --------------                   --------------
  47.0%  25.1Mi .text              25.1Mi  47.3%
  19.4%  10.3Mi .gopclntab         10.3Mi  19.5%
  15.8%  8.44Mi .rodata            8.44Mi  15.9%
   9.9%  5.30Mi .noptrdata         5.30Mi  10.0%
   3.1%  1.68Mi .dynstr            1.68Mi   3.2%
   1.4%   745Ki .eh_frame           745Ki   1.4%
   1.0%   564Ki .dynsym             564Ki   1.0%
   0.7%   364Ki .bss                    0   0.0%
   0.4%   231Ki .data               231Ki   0.4%
   0.3%   183Ki .gcc_except_table   183Ki   0.3%
   0.3%   159Ki .eh_frame_hdr       159Ki   0.3%
   0.3%   158Ki .hash               158Ki   0.3%
   0.1%  62.0Ki .typelink          62.0Ki   0.1%
   0.1%  47.1Ki .gnu.version       47.1Ki   0.1%
   0.1%  35.7Ki .data.rel.ro       35.7Ki   0.1%
   0.0%  22.9Ki .itablink          22.9Ki   0.0%
   0.0%  15.9Ki .noptrbss               0   0.0%
   0.0%  7.21Ki [24 Others]        10.1Ki   0.0%
   0.0%  6.96Ki .rela.plt          6.96Ki   0.0%
   0.0%  4.66Ki .plt               4.66Ki   0.0%
   0.0%       0 [Unmapped]         3.59Ki   0.0%
 100.0%  53.4Mi TOTAL              53.0Mi 100.0%
```

And with RocksDB debug symbols:

```
$ bloaty cockroach-linux-2.6.32-gnu-amd64 
     VM SIZE                       FILE SIZE
 --------------                 --------------
   0.0%       0 .debug_info      73.1Mi  40.6%
  47.0%  25.0Mi .text            25.0Mi  13.9%
   0.0%       0 .debug_loc       18.8Mi  10.5%
   0.0%       0 .debug_str       10.8Mi   6.0%
  19.4%  10.3Mi .gopclntab       10.3Mi   5.7%
  15.8%  8.44Mi .rodata          8.44Mi   4.7%
   0.0%       0 .strtab          6.40Mi   3.6%
   9.9%  5.30Mi .noptrdata       5.30Mi   2.9%
   0.0%       0 .debug_ranges    4.71Mi   2.6%
   0.0%       0 .debug_line      4.26Mi   2.4%
   0.0%       0 .debug_pubtypes  2.40Mi   1.3%
   0.0%       0 .symtab          2.16Mi   1.2%
   0.0%       0 .debug_frame     2.06Mi   1.1%
   3.1%  1.68Mi .dynstr          1.68Mi   0.9%
   0.0%       0 .debug_pubnames  1.21Mi   0.7%
   1.7%   935Ki [38 Others]      1006Ki   0.5%
   0.0%       0 .debug_abbrev     912Ki   0.5%
   1.4%   745Ki .eh_frame         745Ki   0.4%
   1.0%   564Ki .dynsym           564Ki   0.3%
   0.7%   364Ki .bss                  0   0.0%
   0.0%       0 .debug_aranges    280Ki   0.2%
 100.0%  53.3Mi TOTAL             180Mi 100.0%
```

**Comment by tschottdorf** [comment]
Thanks @benesch! Ok, looks like we know where we're at, I'll leave it to @bdarnell to decide whether we want to deal with this size of release binary.

**Comment by bdarnell** [comment]
Nearly doubling the time of an incremental build is a pretty big impact (especially since this cost is not just for `make build`, but also for every test binary). Given that, I'm tempted to say we should build our releases with C++ symbols but leave it off by default for development builds, even though that feels backwards. 

One more thing to try is the `-g1` flag (instead of `-g`, which is equivalent to `-g2`). According to the docs, this includes enough debug information for backtraces, but not enough for debuggers to see local variable information. Since the increase in binary size was far larger than the number of new symbols reported by `nm`, it's possible that the difference is in these local variables that we don't really care about.

**Comment by dt** [comment]
one more variation fwiw: we could build bleeding edge "release" binaries (which we most often use for test clusters) with symbols, but leave them out the true release build (i.e. tagged releases).

**Comment by benesch** [comment]
> One more thing to try is the -g1 flag (instead of -g, which is equivalent to -g2). According to the docs, this includes enough debug information for backtraces, but not enough for debuggers to see local variable information. Since the increase in binary size was far larger than the number of new symbols reported by nm, it's possible that the difference is in these local variables that we don't really care about.

Wow, `-g1` is _way_ better. I'm seeing useful pprof output with only 10MiB of extra debug info.

**Comment by benesch** [comment]
Fixed in #24296.
